docs/syntax-attributes-functions-parameters.md

```diff
@@ -0,0 +1,61 @@
+
+## Design
+
+The following syntax is proposed for attributes:
+
+```ebnf
+parameter-table = '{' '}'
+          | '{' parameter (sep parameter)* '}'
+
+sep = ',' | ';'
+
+parameter = literal 
+          | NAME '=' literal
+
+literal = BOOLEAN | NUMBER | STRING | NIL | parameter-table
+
+list-attribute = NAME parameter-table
+               | NAME
+
+attribute = '@' NAME 
+          | '@[' list-attribute (',' list-attribute)* ']'
+```
+
+In Luau scripts, attributes are specified before the `function` keyword in function definitions. In declaration files, attributes are specified in function type declarations, before the `function` keyword in `declare function (type, ...) : type` syntax, and before the `(` in the `name: (type, ...) : type` syntax. The primary extension proposed to the [Attributes](./syntax-attributes-functions.md) RFC is a new delimited syntax `@[]` for specifying multiple comma-separated attributes with parameters supplied through an optional Luau table. The important features are:
+
+1. Inside the attribute list, `@[]`, attribute names are not allowed to have a leading `@`.
+2. Attribute lists cannot be nested. Attributes cannot be specified on attribute parameters.
+3. Attributes inside `@[]` are separated by a `,`.
+4. Attributes inside `@[]` are allowed to take an optional table of parameters.
+5. The "parameter table" is a Luau table whose entries specify attribute parameters, with or without names, separated by commas or semicolons.
+6. A parameter can be any literal value: `BOOLEAN`, `NUMBER`, `STRING`, `NIL`,  or a `parameter-table`. Attributes can be seen as a way of providing tagged metadata with no evaluation semantics of their own. With that in mind, we only allow these syntactic categories because they evaluate to themselves. This is also the reason we disallow entries of the form `[exp1] = [exp2]` in parameter table, since this would require evaluating `exp1` at construction time. `nil` is allowed for the sake of generality and completeness.
+7. Attributes can be specified before function declarations in multiple ways: `@attr1 @[attr2, attr3{2, "hi"}]`, `@attr1 @attr2 @[attr3{2, "hi"}]`, `@attr1 @[attr3{2, "hi"}] @attr2`, and  `@[attr1, attr2, attr3{2, "hi"}]` are all equivalent.
+8. An attribute with an empty parameter list can be equivalently specified without one, i.e., `@[attr{}]` is same as `@attr`.
+
+The parser is responsible for for enforcing the syntax specified by this RFC and ensuring that attributes are not repeated on a definition or a declaration. The number and type of parameters accepted by a particular attribute will be specified by its own RFC, and enforced by the relevant component of the implementation (typechecker, compiler, etc.), after parsing.
```

Comment:
So if I had

```lua
@unroll("five")
```

The parser is not going to reject this (which is fair enough) but I won't see any errors about it, the compiler has two options: silently ignore `unroll` request, or throw a compilation error which currently does not get shown to the user.

This would mean the linter will have to analyze these attributes and whether they have the correct types so that users get warnings about incorrect arguments.

fwiw, parser rejecting duplicate attributes is not far-fetched. We do this already with table indexers. `{ [A]: B, [C]: D }` is a parse error, it doesn't use the last indexer.